src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -320,17 +336,25 @@ private void EventLoop()
                         {
                             Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
 
-                            Interop.Sys.SocketEvents events = context.HandleSyncEventsSpeculatively(socketEvent.Events);
-                            if (events != Interop.Sys.SocketEvents.None)
+                            if (inlineSocketCompletionsEnabled && context.PreferInlineCompletions)
```

Comment:
Why is the inlineSocketCompletionsEnabled check necessary?  Since we initialize Socket.PreferInlineCompletions based on the environment variable, won't checking it be sufficient?

Reply:
Setting `PreferInlineCompletions` is done by the _developer_ for a `Socket` that is safe to run inline.

Setting `inlineSocketCompletionsEnabled` (aka DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS envvar) is done by the _app user_ to inline completions on Sockets that have `PreferInlineCompletions` set to `true`.

Reply:
@stephentoub does this make sense?

Reply:
> Setting PreferInlineCompletions is done by the developer for a Socket that is safe to run inline.
> Setting inlineSocketCompletionsEnabled (aka DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS envvar) is done by the app user to inline completions on Sockets that have PreferInlineCompletions set to true.

This isn't how I expected we'd do it.  I'd have thought PreferInlineCompletions is set by a developer or by DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS being set, and PreferInlineCompletions is always respected by the engine.  That way, we can experiment with all sockets being in this mode (by setting the env var to true) and ASP.NET can experiment with just its sockets being in this mode (by using reflection to set the property to true).

What additional scenario is enabled by your approach?  Maybe I'm just missing something.

Reply:
> his isn't how I expected we'd do it. I'd have thought PreferInlineCompletions is set by a developer or by DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS being set, and PreferInlineCompletions is always respected by the engine. That way, we can experiment with all sockets being in this mode (by setting the env var to true) and ASP.NET can experiment with just its sockets being in this mode (by using reflection to set the property to true).

That works for experimentation.

> Prior to shipping, we'd either a) make the property public and remove the environment variable, b) delete the property but keep the environment variable, or c) delete both.

I think we need to ship two knobs:
- one that enables the inlining, which scales up the nr of engines
- one that indicates per socket the usage is safe for inlining

ASP.NET Core apps could somehow enable the inlining by default (cfr using garbage gc). And then for their sockets, indicate they are safe to inline.

Should I change to default PreferInlineCompletions to DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS? Or maybe we keep as-is and do API productization discussion later?

Reply:
> I think we need to ship two knobs

Why this configuration, though? If there ends up being a property on Socket, you shouldn't need to set an environment variable for it to work: you opt-in per socket and you're off and running. And if you want to change all sockets, you set the env var, and it changes that default so that unless sockets opt out, they get the new behavior.

> one that enables the inlining, which scales up the nr of engines

Don't we already have an env var now that controls the number of engines?

> Should I change to default PreferInlineCompletions to DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS?

That would be by preference, unless I'm just missing something about the benefit of the extra knob. I simply don't understand its benefit. What does it let me do that I can't otherwise achieve?

Reply:
> Don't we already have an env var now that controls the number of engines?

Yes. And for almost any system it will run with 1 engine.
The DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS envvar changes that default to `ProcessorCount` + allowing inlining.

I'll change the default of PreferInlineCompletions to DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS.

Reply:
This is now a single envvar that:
- sets nr of engines to ProcessorCount
- sets the default of PreferInlineCompletions